### PR TITLE
Add support for deriving readers and writers for enumerations in the Magnolia module

### PR DESCRIPTION
**Pull request by jcazevedo**
Following the discussion on #796, this adds the `deriveEnumerationReader` and `deriveEnumerationWriter` methods for the Magnolia module.

The derived `ConfigReader`s and `ConfigWriter`s follow the behavior of their counterparts in the generic module and may need to be revisited after #787.

With Magnolia, I didn't find a way to guarantee at compile time that we're only trying to derive enumeration readers and writers for sealed traits of objects. However, the cases in which we can derive instances of `EnumerationConfigReaderBuilder` and `EnumerationConfigWriterBuilder` for "incorrect" types look small. We seem to be able to derive it for case classes with empty parameter lists or whose parameter types are also eligible for an instance of those auxiliary type classes. I don't think this is a deal breaker.

### CODE REVIEW

File: tests/src/test/scala/pureconfig/EnumerationsSuite.scala [hunk 1]
```diff
@@ -72,6 +72,9 @@ class EnumerationsSuite extends BaseSuite {
   }
 
   it should "not allow deriving readers, writers and full converters for enumerations encoded as sealed traits whose subclasses are not all case objects" in {
+    sealed trait Entity
+    case class Person(name: String, surname: String) extends Entity
+    case class Place(name: String, lat: Double, lon: Double) extends Entity
```
Comment (jcazevedo) [inline tests/src/test/scala/pureconfig/EnumerationsSuite.scala:77]: This was declared here because these types weren't accessible from this suite, so the derivations below were not compiling for the wrong reasons.

File: modules/magnolia/src/test/scala/pureconfig/module/magnolia/EnumerationsSuite.scala [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+import pureconfig.module.magnolia.semiauto.reader._
+import pureconfig.module.magnolia.semiauto.writer._
+
+class EnumerationsSuite extends BaseSuite {
+
+  sealed trait Color
+  case object RainyBlue extends Color
+  case object SunnyYellow extends Color
+
+  val conf = ConfigFactory.parseString("{ type: person, name: John, surname: Doe }")
```
Comment (ruippeixotog) [inline modules/magnolia/src/test/scala/pureconfig/module/magnolia/EnumerationsSuite.scala]: We don't seem to be using this anywhere.

File: tests/src/test/scala/pureconfig/EnumerationsSuite.scala (hunk 1 above)
Comment (ruippeixotog) [reply tests/src/test/scala/pureconfig/EnumerationsSuite.scala:77]: Good catch! Can we add a message regex to `illTyped` to prevent it from happening again?

File: modules/magnolia/src/main/scala/pureconfig/module/magnolia/EnumerationConfigReaderBuilder.scala [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+object EnumerationConfigReaderBuilder {
+  type Typeclass[A] = EnumerationConfigReaderBuilder[A]
+
+  def combine[A](ctx: CaseClass[EnumerationConfigReaderBuilder, A]): EnumerationConfigReaderBuilder[A] =
+    new EnumerationConfigReaderBuilder[A] {
+      def build(transformName: String => String): ConfigReader[A] =
+        new ConfigReader[A] {
+          def from(cur: ConfigCursor): Result[A] =
+            if (ctx.isObject) Right(ctx.rawConstruct(Seq.empty))
+            else cur.asConfigValue.right.flatMap(v => cur.failed(NoValidCoproductOptionFound(v, Seq.empty)))
```
Comment (ruippeixotog) [inline modules/magnolia/src/main/scala/pureconfig/module/magnolia/EnumerationConfigReaderBuilder.scala]: How are you planning to handle valid coproduct options in errors? I know you mentioned revisiting this after #787; isn't it more convenient to do this now that we're adding a new implementation? Or do you prefer to do it on a separate PR?

File: tests/src/test/scala/pureconfig/EnumerationsSuite.scala [hunk 4]
```diff
@@ -72,6 +72,9 @@ class EnumerationsSuite extends BaseSuite {
   }
 
   it should "not allow deriving readers, writers and full converters for enumerations encoded as sealed traits whose subclasses are not all case objects" in {
```
Comment (ruippeixotog) [inline tests/src/test/scala/pureconfig/EnumerationsSuite.scala:74]: Can we have a test checking that we disallow derivation of enum readers for case classes at compile time, like this one? I'm ok with ignoring the zero-arg case class edge case, but I'd like to ensure we don't generate type classes for case classes with arguments.

File: modules/magnolia/src/main/scala/pureconfig/module/magnolia/EnumerationConfigReaderBuilder.scala (hunk 3 above)
Comment (ruippeixotog) [reply modules/magnolia/src/main/scala/pureconfig/module/magnolia/EnumerationConfigReaderBuilder.scala]: Also, is `NoValidCoproductOptionFound` the right error here?

If I understood the code correctly this is the case where we find ourselves deriving an enum reader for a case class rather than an object. In this case, shouldn't we throw an exception to cause the derivation to fail at compile-time?

**Review by ruippeixotog — CHANGES_REQUESTED** [review]

File: modules/magnolia/src/main/scala/pureconfig/module/magnolia/EnumerationConfigReaderBuilder.scala (hunk 3 above)
Comment (jcazevedo) [reply modules/magnolia/src/main/scala/pureconfig/module/magnolia/EnumerationConfigReaderBuilder.scala]: > Also, is `NoValidCoproductOptionFound` the right error here?
> 
> If I understood the code correctly this is the case where we find ourselves deriving an enum reader for a case class rather than an object. In this case, shouldn't we throw an exception to cause the derivation to fail at compile-time?

Unfortunately, an exception thrown here will only be visible at run time. I don't think there's a straightforward way to fail the compilation at this point. I suppose we can create a custom `FailureReason` for this, but ideally we wouldn't be deriving type classes for case classes.

File: modules/magnolia/src/test/scala/pureconfig/module/magnolia/EnumerationsSuite.scala (hunk 2 above)
Comment (jcazevedo) [reply modules/magnolia/src/test/scala/pureconfig/module/magnolia/EnumerationsSuite.scala]: It's being used in the first example of the spec. I moved its definition to an inner scope to make its use more evident in d805ae7.

File: tests/src/test/scala/pureconfig/EnumerationsSuite.scala (hunk 1 above)
Comment (jcazevedo) [reply tests/src/test/scala/pureconfig/EnumerationsSuite.scala:77]: Added a message regex in fe4e08b. We should add an `@implicitNotFound` for the enumeration helper type classes, but I'll leave that for another PR.

File: tests/src/test/scala/pureconfig/EnumerationsSuite.scala (hunk 4 above)
Comment (jcazevedo) [reply tests/src/test/scala/pureconfig/EnumerationsSuite.scala:74]: Added a copy of this example to pureconfig-magnolia's test suite in b02e5dc.

File: modules/magnolia/src/main/scala/pureconfig/module/magnolia/EnumerationConfigReaderBuilder.scala (hunk 3 above)
Comment (ruippeixotog) [reply modules/magnolia/src/main/scala/pureconfig/module/magnolia/EnumerationConfigReaderBuilder.scala]: That's unexpected :/ I think we really need this error to stand out, as the behavior differs from shapeless and from the compile-time safety of PureConfig in general. I'd probably go for a different error where we can specify a reason, like `CannotConvert`. However, if you still feel we're ok `NoValidCoproductOptionFound` I'm ok with it too.

It's probably worth asking in magnolia how to make this use case possible, too.

**Review by ruippeixotog — APPROVED** [review]
LGTM, please take a look at my comment above.

File: modules/magnolia/src/main/scala/pureconfig/module/magnolia/EnumerationConfigReaderBuilder.scala (hunk 3 above)
Comment (jcazevedo) [reply modules/magnolia/src/main/scala/pureconfig/module/magnolia/EnumerationConfigReaderBuilder.scala]: I opened an issue in Magnolia (propensive/magnolia#267) to see if this use case would be possible. In propensive/magnolia#263, they introduced the possibility to only define the `dispatch` method. I suppose we could have an implementation for `dispatch` only and provide an alternative derivation method for objects. However, I think we'd still need a macro to provide an evidence that a given type is an object.

For the time being, I've changed the failure reason to a `CannotConvert`, so that we can make the underlying issue for the error more explicit.

**Review by ruippeixotog — APPROVED** [review]
Thank you for building this!
